### ISSUE TITLE
Debug Ember-cli generated application.

### ISSUE THREAD

**Issue by ankeshdave** [Microsoft/vscode-chrome-debug#193]
I am working on an EmberJS app created/used using ember-cli.
I am not able to debug from VScode, I can debug easily using Chrome.

Reproduction Steps

1. Create EmberJS application using `ember generate <applicationName>`
2. Build and serve application using `ember serve`
3. Start chrome in debugging mode and open server url (http://localhost:4200)
4. Set breakpoint on app.js (anywhere possible)
5. Attach debugger (`launch.json`)
6. Refresh Page in chrome (Breakpoint does not hit)

Disgnostic log 
[vscode-chrome-debug.txt](https://github.com/Microsoft/vscode-chrome-debug/files/350005/vscode-chrome-debug.txt)

ember-cli-build.js file (Allows to enable source maps, stop minification etc)
[ember-cli-build.txt](https://github.com/Microsoft/vscode-chrome-debug/files/349951/ember-cli-build.txt)

Launch.json
```
{
    "version": "0.2.0",
    "configurations": [{
        "name": "Launch localhost with sourcemaps",
        "type": "chrome",
        "request": "launch",
        "url": "http://localhost:4200",
        "sourceMaps": true,
        "webRoot": "${workspaceRoot}\\dist\\",
        "diagnosticLogging": true
    }, {
        "name": "Attach with sourcemaps",
        "type": "chrome",
        "request": "attach",
        "port": 9222,
        "sourceMaps": true,
        "webRoot": "${workspaceRoot}\\dist\\",
        "diagnosticLogging": true
    }]
}
```

**Comment by ankeshdave** [comment]
Current Workaround i am using :
In VS Code go to the output folder (default is `dist`)
Open file `dist\assets\<applicationName.js>`
Do an **in file search** for `app`
We will see a source code similar to `app.js` (ember-cli is concatenating all js files [dunno how to stop])
Place breakpoints there.
:8ball: It works!! :8ball: :smile: 

But if the project grows big (Current in which i am working has around 50-55 components) it becomes tedious.
you make changes in `app\component\<componentName>.js` the you go and search for the same component in `dist\assets\<applicationName.js>` to set a break point.

**Comment by roblourens** [comment]
The problem is that this extension tries to map sources to actual files on disk, but the paths in the sourcemap that Ember produces aren't quite right. That's fine when you're debugging in Chrome, because the source content is inlined in the sourcemap. But for vscode, we need to either figure out how to fix the paths, like by adding a sourceRoot. I've been messing around with that and I can't figure out how to do that in Ember.

Otherwise, this task will also help - Microsoft/vscode-chrome-debug-core#44 - by providing a way to configure this in your vscode launch config.

**Comment by ankeshdave** [comment]
Great, thanks for looking in to it. 
I'll for now, use the work around.

**Comment by roblourens** [comment]
Hey @ankeshdave, in the latest version, you should have a better workaround now. The `sourceMapPathOverrides` launch config option will let you rewrite the sourcemap paths to their actual locations on disk. From your log, I think this should work for you -

```
"sourceMapPathOverrides": {
     "ember-test/*": "${workspaceRoot}/app/*"
 },
```

There are some more details in the README. Let me know how it goes!

**Comment by ankeshdave** [comment]
Hi @roblourens ,
Thanks for the work around, Finally an IDE to debug Ember-cli :smile: 

But i find that i am not able to set break points properly.
Its skips some lines and puts break point on a different line.

I have added a demo project in [this repository](https://github.com/ankeshdave/vsc-debug-embercli-repo).

Here when we try to set breakpoint at **`Line : 7`-> random-comments.js** it moves to **`Line : 12`**.
But we are able to set break point in the exact location in `Chrome Debugger`

Also in my work project the step-by-step debugging jumps/skip many lines. Adding watch on variables does not seem to work as the line indicated may not be a proper line (Sourcemap to source mapping issue ??)

Behavior is the same when i try on `Ubuntu`
Any help of what i am missing here ??

**Comment by roblourens** [comment]
Thanks for sharing your project @ankeshdave, makes this much easier. There's something not quite right with your setup, but I don't know enough about Ember to tell you what it is. When you open this in Chrome, do you see how the sources are not actually the original ES6 sources? It's the generated content, but broken into individual files and contained in the `sourcesContent` property of the sourcemap.

So when you set a breakpoint in VS Code in that original ES6 source, your project doesn't actually have a valid sourcemap between the original source and the generated source, so it ends up on the wrong line. Since they're pretty similar, you'll often be close :)

I don't know enough about Ember to tell you how to fix it, sorry!

**Comment by ankeshdave** [comment]
@roblourens 
As for the Ember setup, there is nothing custom about it, I have just disbaled Uglification there nothing else  - > `ember-cli-build.js`.

ES6 Transpilation : Ember uses `babel` to transpile ES6 to ES5, and the `Broccoli` pipeline concatenates all  transpiled files (components) in to a single file `dist\assets\vsc-debug-embercli-repo.js`
Typescript compiler also transpiles to ES5 directly but it works in VSCode through this (rite ??).

So is `babel` generating incorrect sourcemaps. ??

**Comment by roblourens** [comment]
Maybe Broccoli is overwriting babel's sourcemaps. But one way or another, the sourcemap doesn't map from ES6 to the concatenated file. You can see it in Chrome devtools yourself.

**Comment by sly7-7** [comment]
I was looking for that but can't find a way of debugging my ember-cli-rails app. For posterity, maybe this could help. Here is the config:
```
{
      "name": "Attach to Chrome, with sourcemaps",
        "type": "chrome",
        "request": "attach",
        "port": 9222,
        "sourceMaps": true,
        "webRoot": "${workspaceRoot}/tmp/ember-cli/apps/frontend",
        "diagnosticLogging": true,
        "sourceMapPathOverrides": {
            "frontend/*": "${workspaceRoot}/frontend/app/*"
        }
 }
```

Hope this helps

**Comment by ahilliker** [comment]
I have just started using VSCode with my ember app, and I'm running into this same issue.  Is it closed because the solution lies in an effort on the ember side to fix their sourcemaps?

I would LOVE to get this working properly.  Debugging ember from within the IDE would make life much simpler.  This would seriously be the killer app for VSCode vs all the other options out there when using ember.

**Comment by roblourens** [comment]
Yes but I vaguely remember hearing that they fixed it some months ago. I have no idea what the status is.

**Comment by ahilliker** [comment]
I'm looking at this:

https://github.com/ember-cli/ember-cli/issues/6066

Which is referenced above, and seeing that there was an option to hopefully fix the sourcemap in ember 2.12-beta which was broken in the 2.13 release.  So it looks like there is progress!

**Comment by dguayrobotiq** [comment]
Any update on this? I can't make it work. I've tried multiple different configurations but they all fail.
